util/topsql/stmtstats/stmtstats.go

```diff
@@ -15,59 +15,134 @@
 	"go.uber.org/atomic"
+	"go.uber.org/zap"
 )
 
 var _ StatementObserver = &StatementStats{}
 
+// nowFunc returns the current time; it's overridden in tests.
+var nowFunc = time.Now
+
 // StatementObserver is an abstract interface as a callback to the corresponding
 // position of TiDB's SQL statement execution process. StatementStats implements
 // StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
 // The caller only needs to be responsible for calling different methods at the
 // corresponding locations, without paying attention to implementation details.
 type StatementObserver interface {
-	// OnExecutionBegin should be called before statement execution.
-	OnExecutionBegin(sqlDigest, planDigest []byte)
+	// OnReceiveCmd should be called on TiDB receive command request.
+	OnReceiveCmd()
+
+	// OnSQLAndPlanDigestFirstReady should be called on statement SQL and plan digest are ready.
+	// This function only use to record the SQL and plan digest, it doesn't matter if you delay calling this function.
+	OnSQLAndPlanDigestFirstReady(sqlDigest, planDigest []byte)
 
-	// OnExecutionFinished should be called after the statement is executed.
-	OnExecutionFinished(sqlDigest, planDigest []byte)
+	// OnExecutionFinished should be called after the statement execution finish.
+	OnExecutionFinished()
 }
 
 // StatementStats is a counter used locally in each session.
 // We can use StatementStats to count data such as "the number of SQL executions",
 // and it is expected that these statistics will eventually be collected and merged
 // in the background.
 type StatementStats struct {
+	seCtx    StatementExecutionContext
+	state    StmtExecState
 	mu       sync.Mutex
 	data     StatementStatsMap
 	finished *atomic.Bool
 }
 
+// StmtExecState is the state of StatementStats
+type StmtExecState int
+
+var (
+	stmtExecStateInitial       StmtExecState = 0
+	stmtExecStateCmdReceived   StmtExecState = 1
+	stmtExecStateDigestIsReady StmtExecState = 2
+	stmtExecStateFinished      StmtExecState = 3
+)
+
+// String implements Stringer interface.
+func (s StmtExecState) String() string {
+	switch s {
+	case stmtExecStateInitial:
+		return "initial"
+	case stmtExecStateCmdReceived:
+		return "cmd_received"
+	case stmtExecStateDigestIsReady:
+		return "digest_is_ready"
+	case stmtExecStateFinished:
+		return "finish"
+	default:
+		return fmt.Sprintf("unknow_%d", int(s))
+	}
+}
+
 // CreateStatementStats try to create and register an StatementStats.
 func CreateStatementStats() *StatementStats {
 	stats := &StatementStats{
+		state:    stmtExecStateInitial,
 		data:     StatementStatsMap{},
 		finished: atomic.NewBool(false),
 	}
 	globalAggregator.register(stats)
 	return stats
 }
 
-// OnExecutionBegin implements StatementObserver.OnExecutionBegin.
-func (s *StatementStats) OnExecutionBegin(sqlDigest, planDigest []byte) {
+// OnReceiveCmd implements StatementObserver.OnParseBegin.
+func (s *StatementStats) OnReceiveCmd() {
```

Comment:
maybe better to check and output invalid state here, for example, it expects state to be initial or finished (but proceed with overridden state)